File: Go/Go_100-999/593.go
// Go
// leetcode 593
// https://leetcode.cn/problems/valid-square/

type Cnt struct{
	Min, c1, Max, c2 int
}

func getCnt() Cnt {
	return Cnt{1000_000_000, 0, -1000_000_000, 0}
}

func (t *Cnt)insert(v int) {
	if v < t.Min {
		t.Min = v
		t.c1 = 1
	} else if v == t.Min {
		t.c1++
	}

	if v > t.Max {
		t.Max = v
		t.c2 = 1
	} else if t.Max == v {
		t.c2++
	}
}

func (t *Cnt) check() bool {
	if t.Min * 2 != t.Max {
		return false
	}

	if t.c1 != 4 {
		return false
	}

	if t.c2 != 2 {
		return false
	}

	return true
}

func getD2(p1, p2 []int) int {
	dx, dy := p1[0] - p2[0], p1[1] - p2[1]
	return dx * dx + dy * dy
}

func validSquare(p1 []int, p2 []int, p3 []int, p4 []int) bool {
	cnt := getCnt()
	
	cnt.insert(getD2(p1, p2))
	cnt.insert(getD2(p1, p3))
	cnt.insert(getD2(p1, p4))
	cnt.insert(getD2(p3, p2))
	cnt.insert(getD2(p4, p2))
	cnt.insert(getD2(p3, p4))

	return cnt.check()
}

File: cpp/cpp_100-999/593.cc
// C++
// leetcode 593
// https://leetcode.cn/problems/valid-square/

class Solution {
public:
	int getDis2(vector<int> &a, vector<int> &b) {
		int dx = a[0] - b[0];
		int dy = a[1] - b[1];
		return dx * dx + dy * dy;
	}

    bool validSquare(vector<int>& p1, vector<int>& p2, vector<int>& p3, vector<int>& p4) {
    	map<int, int> d2;
    	d2[getDis2(p1, p2)]++;
    	d2[getDis2(p1, p3)]++;
    	d2[getDis2(p1, p4)]++;
    	d2[getDis2(p2, p3)]++;
    	d2[getDis2(p2, p4)]++;
    	d2[getDis2(p3, p4)]++;
    	
    	int minD2 = INT_MAX, maxD2 = INT_MIN;
    	for (auto &[k, v] : d2) {
    		minD2 = min(minD2, k);
    		maxD2 = max(maxD2, k);
    	}
    	if (minD2 * 2 != maxD2)
    		return false;
    	if (d2[minD2] != 4)
    		return false;

    	if (d2[maxD2] != 2)
    		return false;

    	return true;
    }
};